File: Lab1.c
#include <stdio.h>
#include <math.h>

/* Function to print IP address from 32-bit integer */
void printIPAddress(unsigned int ipAddress) {
    printf("%u.%u.%u.%u",
           (ipAddress >> 24) & 255,
           (ipAddress >> 16) & 255,
           (ipAddress >> 8) & 255,
           ipAddress & 255);
}

/* Function to count number of 1 bits */
int countOneBits(unsigned int number) {
    int count = 0;
    while (number) {
        count += number & 1;
        number >>= 1;
    }
    return count;
}

int main() {

    unsigned int ipOctet[4];
    unsigned int maskOctet[4];

    unsigned int ipAddress = 0;
    unsigned int subnetMask = 0;

    printf("Enter IP Address: ");
    scanf("%u.%u.%u.%u",
          &ipOctet[0], &ipOctet[1], &ipOctet[2], &ipOctet[3]);

    printf("Enter Subnet Mask: ");
    scanf("%u.%u.%u.%u",
          &maskOctet[0], &maskOctet[1], &maskOctet[2], &maskOctet[3]);

    /* Convert IP and Subnet Mask to 32-bit form */
    for (int i = 0; i < 4; i++) {
        ipAddress = (ipAddress << 8) | ipOctet[i];
        subnetMask = (subnetMask << 8) | maskOctet[i];
    }

    /* Network and Broadcast Address */
    unsigned int networkID = ipAddress & subnetMask;
    unsigned int broadcastID = networkID | (~subnetMask);

    /* Bit calculations */
    int totalNetworkBits = countOneBits(subnetMask);
    int totalHostBits = 32 - totalNetworkBits;

    int addressesPerSubnet = pow(2, totalHostBits);
    int hostsPerSubnet = addressesPerSubnet - 2;

    /* Class identification */
    int defaultClassBits;
    if (ipOctet[0] <= 127)
        defaultClassBits = 8;
    else if (ipOctet[0] <= 191)
        defaultClassBits = 16;
    else
        defaultClassBits = 24;

    int borrowedBits = totalNetworkBits - defaultClassBits;
    int numberOfSubnets = pow(2, borrowedBits);

    printf("\nRESULTS: \n");
    printf("Number of Subnets: %d\n", numberOfSubnets);
    printf("Addresses per Subnet: %d\n", addressesPerSubnet);
    printf("Hosts per Subnet: %d\n", hostsPerSubnet);

    printf("\nNetwork ID: ");
    printIPAddress(networkID);

    printf("\nBroadcast ID: ");
    printIPAddress(broadcastID);

    printf("\n\nSubnet-wise Network and Broadcast IDs:\n");

    unsigned int subnetIncrement = addressesPerSubnet;

    for (int subnetIndex = 0; subnetIndex < numberOfSubnets; subnetIndex++) {

        unsigned int subnetNetworkID =
            networkID + (subnetIndex * subnetIncrement);

        unsigned int subnetBroadcastID =
            subnetNetworkID + subnetIncrement - 1;

        printf("\nSubnet %d\n", subnetIndex + 1);
        printf("Network ID: ");
        printIPAddress(subnetNetworkID);
        printf("\nBroadcast ID: ");
        printIPAddress(subnetBroadcastID);
        printf("\n");
    }

    return 0;
}

File: lab1.java
import java.util.Scanner;

public class SimpleSubnet {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        char ipClass;
        int defaultMask = 0;
        int cidr;
        int[] ip = new int[4];

        System.out.print("Select IP Class (A / B / C): ");
        ipClass = sc.next().charAt(0);

        if (ipClass == 'A' || ipClass == 'a') {
            System.out.println("Class A");
            System.out.println("IP Range: 1.0.0.0 to 126.255.255.255");
            defaultMask = 8;
        }
        else if (ipClass == 'B' || ipClass == 'b') {
            System.out.println("Class B");
            System.out.println("IP Range: 128.0.0.0 to 191.255.255.255");
            defaultMask = 16;
        }
        else if (ipClass == 'C' || ipClass == 'c') {
            System.out.println("Class C");
            System.out.println("IP Range: 192.0.0.0 to 223.255.255.255");
            defaultMask = 24;
        }
        else {
            System.out.println("Invalid Class");
            return;
        }

        System.out.println("Default Subnet Mask: /" + defaultMask);

        System.out.print("Enter IP Address: ");
        String ipAddress = sc.next();
        String[] parts = ipAddress.split("\\.");

        for (int i = 0; i < 4; i++) {
            ip[i] = Integer.parseInt(parts[i]);
        }

        System.out.print("Enter CIDR: ");
        cidr = sc.nextInt();

        if (cidr < defaultMask) {
            System.out.println("Invalid CIDR for selected class!");
            return;
        }

        int hostBits = 32 - cidr;
        int blockSize = (int) Math.pow(2, hostBits);

        int networkID = ip[3] - (ip[3] % blockSize);
        int broadcastID = networkID + blockSize - 1;

        int numberOfSubnets = (int) Math.pow(2, cidr - defaultMask);
        int hostsPerSubnet = blockSize - 2;

        System.out.println("\nNumber of Subnets: " + numberOfSubnets);
        System.out.println("Hosts per Subnet: " + hostsPerSubnet);

        System.out.println("\nNetwork ID: "
                + ip[0] + "." + ip[1] + "." + ip[2] + "." + networkID);

        System.out.println("Broadcast Address: "
                + ip[0] + "." + ip[1] + "." + ip[2] + "." + broadcastID);

        System.out.println("Valid Host Range: "
                + ip[0] + "." + ip[1] + "." + ip[2] + "." + (networkID + 1)
                + " to "
                + ip[0] + "." + ip[1] + "." + ip[2] + "." + (broadcastID - 1));

        sc.close();
    }
}
